compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -3766,69 +3806,100 @@ abstract class GenJSCode extends plugins.PluginComponent
+                  retTpe)
             }
           }, { // else
             callStatement
           })(jstpe.AnyType)
         }
+
+        /* For primitive types, we need to handle two cases. The method could
+         * either be defined in the boxed class of the primitive type, or it
+         * could be defined in the primitive class itself. In the former case,
+         * we generate a direct call to the actual (hijacked) method. In the
+         * latter case, we directly generate the primitive operation.
+         */
+
+        def addCallForPrimitive(primitiveClass: Symbol): Boolean = {
+          val boxedClass = definitions.boxedClass(primitiveClass)
+          val boxedTpe = boxedClass.tpe
+          val methodInBoxedClass = matchingSymIn(boxedClass)
+          if (methodInBoxedClass != NoSymbol && methodInBoxedClass.isPublic) {
+            callStatement = js.If(genIsInstanceOf(callTrg, boxedTpe), {
+              boxIfNeeded(
+                  genApplyMethod(
+                      genAsInstanceOf(callTrg, boxedTpe),
+                      methodInBoxedClass,
+                      arguments),
+                  methodInBoxedClass.tpe.resultType)
+            }, { // else
+              callStatement
+            })(jstpe.AnyType)
+            true
+          } else {
+            val methodInPrimClass = matchingSymIn(primitiveClass)
+            if (methodInPrimClass != NoSymbol && methodInPrimClass.isPublic) {
+              def isIntOrLongKind(kind: TypeKind) = kind match {
+                case CharKind     => false
+                case _:INT | LONG => true
+                case _            => false
+              }
+              val ignoreBecauseItMustBeAnInt = {
+                primitiveClass == DoubleClass &&
+                toTypeKind(methodInPrimClass.tpe.resultType) == DoubleKind &&
+                isIntOrLongKind(toTypeKind(sym.tpe.resultType))
+              }
+              if (ignoreBecauseItMustBeAnInt) {
+                // Fall through to the Int case that will come next
+                false
+              } else {
+                callStatement = js.If(genIsInstanceOf(callTrg, boxedTpe), {
+                  val castCallTrg =
+                    makePrimitiveUnbox(callTrg, primitiveClass.tpe)
+                  val call = genPrimitiveOpForReflectiveCall(methodInPrimClass,
+                      castCallTrg, arguments)
+                  boxIfNeeded(call, methodInPrimClass.tpe.resultType)
+                }, { // else
+                  callStatement
+                })(jstpe.AnyType)
+                true
+              }
+            } else {
+              false
+            }
+          }
+        }
+
+        addCallForPrimitive(BooleanClass)
+        addCallForPrimitive(LongClass)
+        addCallForPrimitive(CharClass)
+
+        /* For primitive numeric types that box as JS numbers, find the first
+         * one that matches. It will be able to handle the subsequent cases.
```

Comment:
`ignoreBecauseItMustBeAnInt` was always there. It handles the case
```scala
def foo(x: Any { def /(x: Int): Int }): Int = x / 2
foo(5)
```
Normally, `x` would be matched as a `Double`, because `Double` has a `def (x: Int)`. However, that would perform a double division and return 2.5 instead of the expected 2. Taking the result type into account allows us to patch that corner case. The same trick allows addition to wrap around 32-bits as it should. It's a very ad hoc case that has been there since https://github.com/scala-js/scala-js/commit/23f4ed3224517a30dea7345ea0fa78f82f83a8a5, i.e., the very day we decided not to box `Int`s anymore.

Generating all cases in order of subtyping check does *not* do the correct thing, for example for `compareTo(Object)`. Also we would still have the int vs double operation issue (e.g., 32-bit wrap around or integer division), only reversed. I don't think this would help in the slightest.

I don't quite understand what you mean by handling operations by their categories. From what I understand, it doesn't seem to simplify anything, only making the code longer by having to deal with the categories.